File: tests/ssurjano_benchmark/permdb.py
import numpy as np

def permdb(xx, b=0.5):
    ##########################################################################
    #URRIN ET AL. (2013) PERM FUNCTION d, beta\n
    #Authors: Sonja Surjanovic, Simon Fraser University\n
    #          Derek Bingham, Simon Fraser University\n
    #Questions/Comments: Please email Derek Bingham at dbingham@stat.sfu.ca.\n
    ####################################################################
    
    xx = np.array(xx)
    b = b
    
    n = len(xx)
    ii = np.arange(1, n+1)
    jj = np.tile(ii, (n, n))
    
    xxmat = np.tile(xx, (n, n))
    inner = np.sum((jj**ii + b)*((xxmat/jj)**ii - 1), axis=0)
    outer = np.sum(inner**2)
    
    y = outer
    return(y)

File: tests/ssurjano_benchmark/permdb.r
permdb <- function(xx, b=0.5)
{
  ##########################################################################
  #
  # PERM FUNCTION d, beta
  #
  # Authors: Sonja Surjanovic, Simon Fraser University
  #          Derek Bingham, Simon Fraser University
  # Questions/Comments: Please email Derek Bingham at dbingham@stat.sfu.ca.
  #
  # Copyright 2013. Derek Bingham, Simon Fraser University.
  #
  # THERE IS NO WARRANTY, EXPRESS OR IMPLIED. WE DO NOT ASSUME ANY LIABILITY
  # FOR THE USE OF THIS SOFTWARE.  If software is modified to produce
  # derivative works, such modified software should be clearly marked.
  # Additionally, this program is free software; you can redistribute it 
  # and/or modify it under the terms of the GNU General Public License as 
  # published by the Free Software Foundation; version 2.0 of the License. 
  # Accordingly, this program is distributed in the hope that it will be 
  # useful, but WITHOUT ANY WARRANTY; without even the implied warranty 
  # of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU 
  # General Public License for more details.
  #
  # For function details and reference information, see:
  # http://www.sfu.ca/~ssurjano/
  #
  ##########################################################################
  #
  # INPUTS:
  #
  # xx = c(x1, x2)
  # b  = constant (optional), with default value 0.5
  #
  ##########################################################################
  
  d <- length(xx)
  ii <- c(1:d)
  jj <- matrix(rep(ii,times=d), d, d, byrow=TRUE)

  xxmat <- matrix(rep(xx,times=d), d, d, byrow=TRUE)
  inner <- rowSums((jj^ii+b)*((xxmat/jj)^ii-1))	
  outer <- sum(inner^2)
	
  y <- outer
  return(y)
}
